encoding/ccf/decode.go

```diff
@@ -0,0 +1,2039 @@
+
+	common.UseMemory(d.gauge, common.MemoryUsage{
+		Kind: common.MemoryKindRawString,
+		// No need to add 1 to account for empty string: string is metered in Path struct.
+		Amount: uint64(len(domain)),
+	})
+
+	// Decode identifier.
+	identifier, err := d.dec.DecodeString()
+	if err != nil {
+		return nil, err
+	}
+
+	common.UseMemory(d.gauge, common.MemoryUsage{
+		Kind: common.MemoryKindRawString,
+		// No need to add 1 to account for empty string: string is metered in Path struct.
+		Amount: uint64(len(identifier)),
+	})
+
+	return cadence.NewMeteredPath(d.gauge, domain, identifier), nil
+}
+
+// decodeCapability decodes encoded capability-value as
+// language=CDDL
+// capability-value = [
+//
+//	address: address-value,
+//	path: path-value
+//
+// ]
+func (d *Decoder) decodeCapability(typ *cadence.CapabilityType, types *cadenceTypeByCCFTypeID) (cadence.Value, error) {
+	// typ can be different from runtime CapabilityType because borrow type can be nil.
+	// In this case, runtime type is encoded with the value (as tag content for tag CBORTagTypeAndValue).
+
+	// Check next encoded CBOR type.
+	nextType, err := d.dec.NextType()
+	if err != nil {
+		return nil, err
+	}
+
+	if nextType == cbor.TagType {
+		err := decodeCBORTagWithKnownNumber(d.dec, CBORTagTypeAndValue)
+		if err != nil {
+			return nil, fmt.Errorf("unexpected encoded value of Cadence type %s (%T): %s", typ.ID(), typ, err.Error())
+		}
+
+		// Decode ccf-type-and-value-message.
+		return d.decodeTypeAndValue(types)
+	}
+
+	// Decode array head of length 2.
+	err = decodeCBORArrayWithKnownSize(d.dec, 2)
+	if err != nil {
+		return nil, err
+	}
+
+	// Decode address.
+	address, err := d.decodeAddress()
+	if err != nil {
+		return nil, err
+	}
+
+	// Decode path.
+	path, err := d.decodePath()
+	if err != nil {
+		return nil, err
+	}
+
+	return cadence.NewMeteredStorageCapability(
+		d.gauge,
+		path.(cadence.Path),
+		address.(cadence.Address),
+		typ.BorrowType), nil
+}
+
+// decodeNullableTypeValue decodes encoded type-value / nil.
+// See _decodeNullableTypeValue() for details.
+func (d *Decoder) decodeNullableTypeValue() (cadence.Value, error) {
+	t, err := d._decodeNullableTypeValue(newCadenceTypeByCCFTypeID())
+	if err != nil {
+		return nil, err
+	}
+	return cadence.NewMeteredTypeValue(d.gauge, t), nil
+}
+
+// _decodeTypeValue decodes encoded type-value as
+// language=CDDL
+// type-value = simple-type-value
+//
+//	/ optional-type-value
+//	/ varsized-array-type-value
+//	/ constsized-array-type-value
+//	/ dict-type-value
+//	/ struct-type-value
+//	/ resource-type-value
+//	/ contract-type-value
+//	/ event-type-value
+//	/ enum-type-value
+//	/ struct-interface-type-value
+//	/ resource-interface-type-value
+//	/ contract-interface-type-value
+//	/ function-type-value
+//	/ reference-type-value
+//	/ restricted-type-value
+//	/ capability-type-value
+//	/ type-value-ref
+func (d *Decoder) _decodeTypeValue(visited *cadenceTypeByCCFTypeID) (cadence.Type, error) {
```

Comment:
Why does this function and `_decodeNullableTypeValue` have an underscore prefix?

Reply:
The helper function (with underscore prefix) was called by function with non-prefixed name.

I simplified this in commit dbbbbd9927652d2c628a05868fee0f7db865d205 by inlining.